enterprise_access/apps/provisioning/models.py

```diff
@@ -0,0 +1,191 @@
+from django_countries import countries
+
+from enterprise_access.apps.workflow.exceptions import UnitOfWorkException
+from enterprise_access.apps.workflow.models import AbstractWorkflow, AbstractWorkflowStep
+from enterprise_access.apps.workflow.serialization import BaseInputOutput
+
+from .api import get_or_create_enterprise_admin_users, get_or_create_enterprise_customer
+
+
+@define
+class GetCreateCustomerStepInput(BaseInputOutput):
+    """
+    The input object to be used for the business logic of get-or-creating
+    an EnterpriseCustomer.
+    """
+    KEY = 'create_customer_input'
+
+    name: str = field(validator=validators.instance_of(str))
+    slug: str = field(validator=validators.instance_of(str))
+    country: str = field(validator=validators.in_(dict(countries)))
+
+
+@define
+class GetCreateCustomerStepOutput(BaseInputOutput):
+    """
+    The output object that stores the result of get-or-creating
+    an EnterpriseCustomer.
+    """
+    KEY = 'create_customer_output'
+
+    uuid: UUID = field(validator=validators.instance_of(UUID))
+    name: str = field(validator=validators.instance_of(str))
+    slug: str = field(validator=validators.instance_of(str))
+    country: str = field(validator=validators.in_(dict(countries)))
+
+
+class CreateCustomerStepException(UnitOfWorkException):
+    """
+    Exception raised when an EnterpriseCustomer could not be created or fetched.
+    """
+
+
+class GetCreateCustomerStep(AbstractWorkflowStep):
+    """
+    Workflow step for creating a new customer, or returning an existing record
+    based on matching title value.
+    """
+    exception_class = CreateCustomerStepException
+    input_class = GetCreateCustomerStepInput
+    output_class = GetCreateCustomerStepOutput
+
+    def process_input(self, accumulated_output=None, **kwargs):
+        """
+        Gets or creates an enterprise customer record.
+
+        Params:
+          accumulated_output (obj): An optional accumulator object to which
+            the resulting output can be added.
+
+        Returns:
+          An instance of ``self.output_class``.
+        """
+        input_object = self.input_object
+        result_dict = get_or_create_enterprise_customer(
+            name=input_object.name,
+            slug=input_object.slug,
+            country=input_object.country,
+        )
+        return self.output_class.from_dict(result_dict)
+
+    def get_workflow_record(self):
+        return ProvisionNewCustomerWorkflow.objects.filter(
+            uuid=self.workflow_record_uuid,
+        ).first()
+
+    def get_preceding_step_record(self):
+        return None
+
+
+@define
+class GetCreateEnterpriseAdminUsersInput(BaseInputOutput):
+    """
+    The input object to be used for the fetching or creating
+    of enterprise admin users.
+    """
+    KEY = 'create_enterprise_admin_users_input'
+
+    # enterprise_customer_uuid: UUID = field(validator=validators.instance_of(UUID))
```

Comment:
[question] if this is no longer needed, should it be removed? Or does it need to be uncommented?